compiler/noirc_frontend/src/monomorphization/debug_types.rs

```diff
@@ -0,0 +1,70 @@
+use crate::hir_def::types::Type;
+pub use noirc_errors::debug_info::{Types, VariableTypes, Variables};
+use noirc_printable_type::PrintableType;
+use std::collections::HashMap;
+
+/// We keep a collection of the debug variables and their types in this
+/// structure. The fe_var_id refers to the ID given when inserting the
+/// instrumentation probe. This variable does not have a type yet and hence it
+/// can be instantiated multiple types if it's in the context of a generic
+/// variable. The var_id refers to the ID of the instantiated variable which
+/// will have a valid type.
+#[derive(Debug, Clone, Default)]
+pub struct DebugTypes {
+    fe_to_vars: HashMap<u32, u32>,          // fe_var_id => var_id
+    variables: HashMap<u32, (String, u32)>, // var_id => (var_name, type_id)
+    types: HashMap<PrintableType, u32>,
+    id_to_type: HashMap<u32, PrintableType>,
+    next_type_id: u32,
+    next_var_id: u32,
+}
+
+impl DebugTypes {
+    pub fn insert_var(&mut self, fe_var_id: u32, var_name: &str, var_type: Type) -> u32 {
+        let ptype: PrintableType = var_type.follow_bindings().into();
+        let type_id = self.types.get(&ptype).cloned().unwrap_or_else(|| {
+            let type_id = self.next_type_id;
+            self.next_type_id += 1;
+            self.types.insert(ptype.clone(), type_id);
+            self.id_to_type.insert(type_id, ptype);
+            type_id
+        });
+        let existing_var_id = self.fe_to_vars.get(&fe_var_id).and_then(|var_id| {
+            let (_, existing_type_id) = self.variables.get(var_id).unwrap();
```

Comment:
Is this guaranteed `var_id` is present in this hash map? I can't seem to tell from the function

Reply:
Yes, it's guaranteed because if the `fe_var_id` is found, it means it's already being mapped to at least one specific (ie. non-generic) data type and instantiated in the tracked variables map. The code for this insertion is in the else part at the end of this method.

I do agree it's confusing. I'll replace the `u32` types with more specific ones and re-evaluate the resulting code.

There is also one "hidden" assumption in all this code which I don't think we documented properly. Given an `fe_var_id` we don't always have access to the underlying type (eg. for dropping them), and so we don't know if we're using the "correct" `var_id`. The assumption is that all references to the variable are local to the function being monomorphized and they happen *after* the variable is declared or assigned to, and we *do* have the type at that precise moment. When that happens we either reuse the `var_id` or we create a fresh new one, which should be correct for the rest of the scope being compiled. TBH, we didn't initially take into account generic functions, so we introduced the mapping from `fe_var_id` into `var_id` later in the development.

Reply:
I added several comments in the code explaining the difference between the `source_var_id` and the `var_id`. There's also the description of the `source_to_debug_vars` field in `DebugTypeTracker` that explains why the mapping is necessary.